### PR TITLE
More governor code review rework

### PR DESCRIPTION
**Pull request by bruce-riley**
Change-Id: I0ae094b16a8e0469f83f711e7118936abf70b109

- - -
This change is [Reviewable](https://reviewable.io/reviews/certusone/wormhole/1404)

### CODE REVIEW

File: node/pkg/governor/governor_prices.go [hunk 1]
```diff
@@ -2,7 +2,7 @@
 //
 // The initial prices are read from the static config (tokens.go). After that, prices are
 // queried from CoinGecko. The chain governor then uses the maximum of the static price and
-// the latest CoinGecko price.
+// the latest CoinGecko price. CoinGecko is polled every five minutes.
```
Comment (SEJeff) [inline node/pkg/governor/governor_prices.go:5]: If you made the polling interval  a constant this comment wouldn't be necessary. Maybe?

```
const coinGeckoQueryIntervalMins = 5
```

File: node/pkg/governor/governor_prices.go [hunk 2]
```diff
@@ -38,31 +40,32 @@ func (gov *ChainGovernor) initCoinGecko(ctx context.Context) error {
...
+		case <-ctx.Done():
+			return nil
+		case <-timer.C:
+			gov.queryCoinGecko()
+			timer = time.NewTimer(time.Duration(5) * time.Minute)
 		}
-	}()
+	}
 
 	return nil
```
Comment (SEJeff) [inline node/pkg/governor/governor_prices.go]: The linter has a point here, it is impossible here to exit the loop, so no error will be returned.

**Review by SEJeff — DISMISSED** [review]

File: node/pkg/governor/governor_prices.go [hunk 3]
```diff
@@ -38,72 +40,82 @@ func (gov *ChainGovernor) initCoinGecko(ctx context.Context) error {
...
-	}()
 
-	return nil
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-timer.C:
+			gov.queryCoinGecko()
+			timer = time.NewTimer(time.Duration(5) * time.Minute)
```
Comment (hendrikhofstadt) [inline node/pkg/governor/governor_prices.go:66]: Any reason to not use a ticker instead?

**Review by hendrikhofstadt — DISMISSED** [review]

File: node/pkg/governor/governor_prices.go [hunk 4]
```diff
@@ -118,10 +130,45 @@ func (gov *ChainGovernor) queryCoinGecko() {
 				zap.Stringer("cfgPrice", te.cfgPrice),
 				zap.Stringer("coinGeckoPrice", te.coinGeckoPrice),
 			)
+
+			delete(localTokenMap, coinGeckoId)
+		} else {
+			gov.logger.Error("cgov: received a CoinGecko response for an unexpected symbol", zap.String("coinGeckoId", te.coinGeckoId))
```
Comment (bruce-riley) [inline node/pkg/governor/governor_prices.go:136]: This is a nil pointer reference!

**Review by hendrikhofstadt — DISMISSED** [review]

**Review by claudijd — APPROVED** [review]

File: node/pkg/governor/governor.go [hunk 5]
```diff
@@ -375,61 +361,59 @@ func (gov *ChainGovernor) CheckPendingForTime(now time.Time) ([]*common.MessageP
...
-					return msgsToPublish, err
+					gov.logger.Error("cgov: failed to compute value for pending vaa",
+						zap.Stringer("amount", pe.amount),
+						zap.Stringer("price", pe.token.price),
+						zap.String("msgID", pe.msg.MessageIDString()),
+						zap.Error(err),
+					)
+
+					gov.msgsToPublish = msgsToPublish
+					return nil, err
```
[the remark this answers is not in the document]
Comment (bruce-riley) [reply node/pkg/governor/governor.go:381]: Correct, if anything goes wrong, it will throw it back on the list and try again next interval.

**Review by kcsongor — APPROVED** [review]

**Review by claudijd — APPROVED** [review]

**Pull request merged**